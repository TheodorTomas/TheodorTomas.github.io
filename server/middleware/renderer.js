import React from 'react';
import { renderToString } from 'react-dom/server';
import { ServerStyleSheet } from 'styled-components';
import App from '../../src/containers/App'; // import our main App component
const path = require('path');
const fs = require('fs');
export default (req, res) => {
  // point to the html file created by CRA's build tool
  const filePath = path.resolve(__dirname, '..', '..', 'build', 'index.html');
  fs.readFile(filePath, 'utf8', (err, htmlData) => {
    if (err) {
      console.error('err', err); // eslint-disable-line
      return res.status(404).end();
    }
    const sheet = new ServerStyleSheet();
    // render the app as a string
    const html = renderToString(sheet.collectStyles(<App />));
    const css = sheet.getStyleTags();
    // inject the rendered app into our html and send it
    return res.send(
      htmlData
        .replace(
          '<style ssr-styles></style>',
          css
        )
        .replace(
          '<div id="root"></div>',
          `<div id="root">${html}</div>`
        )
    );
  });
};
