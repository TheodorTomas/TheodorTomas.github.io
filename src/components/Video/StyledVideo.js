import styled from 'styled-components';
import { COLORS } from '../../containers/App/theme';

export const StyledVideo = styled.video`
  /* Sizing */
  width: 150px;
  height: 150px;
  margin: auto;
  
  /* Misc */
  background-color: ${COLORS.videoBgColor.hex};
  overflow: hidden;
  transition: border-color 0.25s ease-in-out, transform 0.25s ease-in-out;
  border-radius: 50%;
  border: 2px solid;
  border-color: ${COLORS.darkGrey.rgba(0.5)};
  box-shadow: 2px 2px 2px 1px rgba(0, 0, 0, 0.2);
  
  ${({ isClickable }) => (
    isClickable
      ? `
        &:hover {
          transform: scale(1.01);
          cursor: pointer;
          border-color: ${COLORS.blueGrey.rgba(0.5)};
        }
        `
      : ''
  )}
`;
