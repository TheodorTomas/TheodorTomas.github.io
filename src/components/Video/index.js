/**
 * Video
 *
 * Renders video element which rewinds to start and replays upon click.
 */
import React from 'react';
import PropTypes from 'prop-types';
import StyledVideo from './StyledVideo';

export default class Video extends React.PureComponent {
  video = React.createRef();

  static propTypes = {
    vidSrc: PropTypes.string.isRequired,
  };

  state = {
    isClickable: false,
  };

  componentDidMount() {
    this.handleMount();
  }

  componentWillUnmount() {
    this.handleUnmount();
  }

  rewindVideo = () => {
    if (this.video.current.currentTime <= 0.5) {
      clearInterval(this.intervalRewind);
      this.video.current.addEventListener('ended', this.handleVideoEnding, false);
      this.video.current.play();
    } else if (this.video.current.currentTime <= 1) {
      this.video.current.currentTime += -0.11;
    } else {
      this.video.current.currentTime += -0.33;
    }
  };

  handleVideoEnding = () => {
    this.setState({ isClickable: true });
  };

  handleVideoClick = () => {
    const { isClickable } = this.state;

    if (isClickable) {
      this.setState({ isClickable: false });
      this.rewindVideo();
      this.intervalRewind = setInterval(this.rewindVideo, 66);
    }
  };

  handleMount() {
    if (this.video.current) {
      this.video.current.playbackRate = 2.5;
      this.video.current.addEventListener('ended', this.handleVideoEnding, false);
    }
  }

  handleUnmount() {
    this.video.current.removeEventListener('ended', this.handleVideoEnding);
    clearInterval(this.intervalRewind);
  }

  render() {
    const { vidSrc } = this.props;
    const { isClickable } = this.state;

    return (
      <StyledVideo
        ref={this.video}
        autoPlay
        muted
        title={isClickable ? 'Click me!' : ''}
        isClickable={isClickable}
        onClick={this.handleVideoClick}
      >
        <source src={vidSrc} type="video/mp4" />
      </StyledVideo>
    );
  }
}
