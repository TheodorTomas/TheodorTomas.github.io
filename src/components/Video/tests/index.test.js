import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import Video from '../index';
import { StyledVideo } from '../StyledVideo';

jest.useFakeTimers();

describe('<Video />', () => {
  const defaultProps = {
    vidSrc: 'test.png',
  };
  const renderComponent = (props = defaultProps) => renderer.create(<Video {...props} />);
  const mountComponent = (props = defaultProps) => mount(<Video {...props} />);

  describe('componentDidMount()', () => {
    it('should set the video parameter on mount', () => {
      const component = mountComponent();
      expect(component.instance().video.current).toBeTruthy();
    });

    it('should mount with state matching snapshot', () => {
      const component = mountComponent();
      expect(component.instance().state).toMatchSnapshot();
    });

    it('should call handleMount when mounting', () => {
      const spy = jest.spyOn(Video.prototype, 'handleMount');
      mountComponent();
      expect(spy).toHaveBeenCalled();
    });
  });

  describe('componentWillUnmount()', () => {
    it('should call handleUnmount when unmounting', () => {
      const spy = jest.spyOn(Video.prototype, 'handleUnmount');
      const component = mountComponent();
      component.unmount();
      expect(spy).toHaveBeenCalled();
    });
  });

  describe('handleMount()', () => {
    it('should do nothing if no video.current', () => {
      const component = mountComponent();
      const mockAddEventListener = jest.fn();
      component.instance().video.current = undefined;
      component.instance().handleMount();
      expect(mockAddEventListener).not.toHaveBeenCalled();
    });

    it('should set the video playbackRate correctly', () => {
      const component = mountComponent();
      component.instance().video.current.playbackRate = 0;
      component.instance().handleMount();
      expect(component.instance().video.current.playbackRate).toBe(2.5);
    });

    it('should set the an event listener on the video.current correctly', () => {
      const component = mountComponent();
      const mockAddEventListener = jest.fn();
      component.instance().video.current.addEventListener = mockAddEventListener;
      component.instance().handleMount();
      expect(mockAddEventListener).toHaveBeenCalledWith('ended', component.instance().handleVideoEnding, false);
    });
  });

  describe('handleUnmount()', () => {
    it('should remove the event listener from the video.current on unmount', () => {
      const component = mountComponent();
      const mockRemoveEventListener = jest.fn();
      component.instance().video.current.removeEventListener = mockRemoveEventListener;
      component.instance().handleUnmount();
      expect(mockRemoveEventListener).toHaveBeenCalledWith('ended', component.instance().handleVideoEnding);
      expect(clearInterval).toHaveBeenCalledWith(component.instance().intervalRewind);
    });
  });

  describe('handleVideoEnding()', () => {
    it('should set state to isClickable = false', () => {
      const component = mountComponent();
      const spy = jest.spyOn(Video.prototype, 'setState');
      component.instance().handleVideoEnding();
      expect(spy).toHaveBeenCalledWith({ isClickable: true });
    });
  });

  describe('handleVideoClick()', () => {
    it('should do nothing if state.isClickable = false ', () => {
      const component = mountComponent();
      component.instance().video.current.play = () => {};
      component.instance().setState({ isClickable: false });
      component.update();

      expect(component.instance().state.isClickable).toBe(false);
      component.instance().handleVideoClick();
      expect(component.instance().state.isClickable).toBe(false);
    });

    it('should handle if state.isClickable = true correctly', () => {
      const component = mountComponent();
      component.instance().video.current.play = () => {};
      const setStateSpy = jest.spyOn(Video.prototype, 'setState');
      const rewindVideoSpy = jest.spyOn(component.instance(), 'rewindVideo');

      component.instance().setState({ isClickable: true });
      component.update();
      component.instance().handleVideoClick();

      expect(setStateSpy).toHaveBeenCalled();
      expect(rewindVideoSpy).toHaveBeenCalled();
      expect(setInterval).toHaveBeenCalledWith(component.instance().rewindVideo, 66);
    });
  });

  describe('rewindVideo()', () => {
    it('should handle the instance of the time being less or equal to 0.5', () => {
      const component = mountComponent();
      component.instance().video.current.currentTime = 0;
      const mockAddEventListener = jest.fn();
      const mockPlay = jest.fn();
      component.instance().video.current.addEventListener = mockAddEventListener;
      component.instance().video.current.play = mockPlay;
      component.instance().rewindVideo();
      expect(clearInterval).toHaveBeenCalledWith(component.instance().intervalRewind);
      expect(mockAddEventListener).toHaveBeenCalledWith('ended', expect.any(Function), false);
      expect(mockPlay).toHaveBeenCalled();
    });

    it('should handle the instance of the time being between or equal to 0.5 and 1', () => {
      const component = mountComponent();
      const currentTime = 1;
      component.instance().video.current.currentTime = currentTime;
      component.instance().video.current.play = () => {};
      component.instance().rewindVideo();
      expect(component.instance().video.current.currentTime).toBe(currentTime - 0.11);
    });

    it('should handle the instance of the time being greater than 1', () => {
      const component = mountComponent();
      component.instance().video.current.currentTime = 1.33;
      component.instance().video.current.play = () => {};
      component.instance().rewindVideo();
      expect(component.instance().video.current.currentTime).toBe(1);
    });
  });

  describe('render()', () => {
    it('should render component matching the snapshot', () => {
      const component = renderComponent();
      const tree = component.toJSON();

      expect(tree).toMatchSnapshot();
    });
    it('should render the correct title according to isClickable', () => {
      const component = mountComponent();

      expect(component.instance().state.isClickable).toBe(false);
      expect(component.find(StyledVideo).props().title).toBe('');

      component.instance().setState({ isClickable: true });
      component.update();

      expect(component.find(StyledVideo).props().title).toBe('Click me!');
    });
  });
});
