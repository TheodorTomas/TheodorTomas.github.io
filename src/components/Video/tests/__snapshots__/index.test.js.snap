// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Video /> componentDidMount() should mount with state matching snapshot 1`] = `
Object {
  "isClickable": false,
}
`;

exports[`<Video /> render() should render component matching the snapshot 1`] = `
.c0 {
  width: 150px;
  height: 150px;
  margin: auto;
  background-color: #8f8c86;
  overflow: hidden;
  -webkit-transition: border-color 0.25s ease-in-out,-webkit-transform 0.25s ease-in-out;
  -webkit-transition: border-color 0.25s ease-in-out,transform 0.25s ease-in-out;
  transition: border-color 0.25s ease-in-out,transform 0.25s ease-in-out;
  border-radius: 50%;
  border: 2px solid;
  border-color: rgba(74,74,74,0.5);
  box-shadow: 2px 2px 2px 1px rgba(0,0,0,0.2);
}

<video
  autoPlay={true}
  className="c0"
  muted={true}
  onClick={[Function]}
  title=""
>
  <source
    src="test.png"
    type="video/mp4"
  />
</video>
`;
